File: packages/search/src/heading-slug.ts
/** 見出しテキスト → アンカー id 用 slug（ページ内重複は連番）。 */

export function slugifyHeading(text: string): string {
  const base = text
    .trim()
    .toLowerCase()
    .replace(/[*_`~]/g, "")
    .replace(/\s+/g, "-")
    .replace(/[^\w぀-ヿ㐀-鿿豈-﫿ｦ-ﾟ-]/g, "")
    .replace(/-+/g, "-")
    .replace(/^-+|-+$/g, "");
  return base.length > 0 ? base : "section";
}

export class SlugLedger {
  private readonly used = new Map<string, number>();

  next(text: string): string {
    const base = slugifyHeading(text);
    const count = this.used.get(base) ?? 0;
    this.used.set(base, count + 1);
    return count === 0 ? base : `${base}-${count + 1}`;
  }
}

File: rust/sorane-astro-backend/src/heading_slug.rs
use regex::Regex;
use std::sync::LazyLock;

static SLUG_STRIP_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"[^\w぀-ヿ㐀-鿿豈-﫿ｦ-ﾟ-]").expect("slug strip re"));

pub fn slugify_heading(text: &str) -> String {
    let mut base = text.trim().to_lowercase();
    for ch in ['*', '_', '`', '~'] {
        base = base.replace(ch, "");
    }
    base = base.split_whitespace().collect::<Vec<_>>().join("-");
    base = SLUG_STRIP_RE.replace_all(&base, "").to_string();
    while base.contains("--") {
        base = base.replace("--", "-");
    }
    base = base.trim_matches('-').to_string();
    if base.is_empty() {
        "section".to_string()
    } else {
        base
    }
}

pub struct SlugLedger {
    used: std::collections::HashMap<String, usize>,
}

impl SlugLedger {
    pub fn new() -> Self {
        Self {
            used: std::collections::HashMap::new(),
        }
    }

    pub fn next(&mut self, text: &str) -> String {
        let base = slugify_heading(text);
        let count = *self.used.get(&base).unwrap_or(&0);
        self.used.insert(base.clone(), count + 1);
        if count == 0 {
            base
        } else {
            format!("{base}-{}", count + 1)
        }
    }
}

impl Default for SlugLedger {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deduplicates_headings() {
        let mut ledger = SlugLedger::new();
        assert_eq!(ledger.next("Hello"), "hello");
        assert_eq!(ledger.next("Hello"), "hello-2");
    }
}